# Multiple Regression

1. Here we have multiple independent variables and single dependent variable 
2. Dummy variables are required when we have a categorical data also whenever we have a categorical data then omit one dummy variable to get the accurate predictions
3. Statistical significance of the model is very important with the null hypothesis being set to critical point 5% is more common but 1% used in medical field 

Building a Model
 we cant take each and every variable in our model for prediction as- garbage in gives garbage out and many variables also makes it difficult to explain it to the employers their significance

 5 methods for building a model:
 1. All In - Every independent variable is included in the model and should be considered only when you have prior knowledge or you boss said you or org<br>


 2. Backward elimination<br>
   1.Select the significance level (suppose SL=0.05)<br>
   2.fit the full model woth all possible predictors<br>
   3.consider the predictor wothe the highest P-value. If P>SL then go to step 4 otherwise go to fin(finished)<br>
   4.Remove the predictor<br>
   5.Fit the model without this variable then again go to step 3 until finish.<br>


 3. Forward Selection<br>
   1.select a significance level to enter the model eg, SL=0.05<br>
   2.Fit all the simple regression models y ~ xn. Select the one with the lowest P-value<br>
   3.Keep this Variable and fit all the possible models with one extra predictor added to the one you already have<br>
   4.consider the predictor with the lowest P-value. If P < SL then go to step 3. otherwise finnish<br>


 4. Bidirectional Elimination
   1.select SL<br>
   2.perform the next step of the forward selection (new variable must have: P < SL to enter)<br>
   3.Perform ALL steps of Backward elimation (old variables must have P < SL to stay)<br>
   4.No new variables can enter and no old variables can exit.<br>

 5. All possible models (score comparison)<br>
   1.Select a criterion of goodness of fit (eg. Akaike criterion)<br>
   2.Construct All possible regression Models: 2^n-1 total combinations<br>
   3.Select the one with the best criterion<br>
   

Import Libraries


In [16]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt 


In [17]:
df = pd.read_csv('50_Startups.csv')
#R&D Spend,Administration,Marketing Spend,State,Profit
X = df.drop('Profit',axis=1).values
y = df["Profit"].values
#print(X)
# df = pd.read_csv('50_Startups.csv')
# X = df.iloc[:, :-1].values
# y = df.iloc[:, -1].values

X4 is categorical so we use OneHotEncoding 

In [18]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder' , OneHotEncoder() , [3])], remainder='passthrough')
X = np.array(ct.fit_transform(X))
#print(X)

splitting the data set


In [19]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2,random_state = 0)

No such need of Feature scaling as there are coefficient to compensate
<br>No need to watch for linear regression Assumption<br>
No need to focus on Dummy variables they are handled automatically by the advanced python libraries and classes 

In [20]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

LinearRegression()

Predicting the Test set results (as plotting the multiple regression is not possible for multiple variable so we define two vectors)

In [21]:
y_pred = regressor.predict(X_test)
# print(y_pred.shape)
# print(y_test.shape)
np.set_printoptions(precision=2)
prediction = np.concatenate((y_pred.reshape(len(y_pred), 1), y_test.reshape(len(y_test), 1)), axis=1)
print(prediction)

[[103015.2  103282.38]
 [132582.28 144259.4 ]
 [132447.74 146121.95]
 [ 71976.1   77798.83]
 [178537.48 191050.39]
 [116161.24 105008.31]
 [ 67851.69  81229.06]
 [ 98791.73  97483.56]
 [113969.44 110352.25]
 [167921.07 166187.94]]


Making a single prediction for R&D Spend = 160000, Administration Spend = 130000, Marketing Spend = 300000 and State = 'California'

In [22]:
print(regressor.predict([[1,0,0,160000,130000, 300000]]))

[181566.92]


###### Important note 1: Notice that the values of the features were all input in a double pair of square brackets. That's because the "predict" method always expects a 2D array as the format of its inputs. And putting our values into a double pair of square brackets makes the input exactly a 2D array. Simply put:

###### 1,0,0,160000,130000,300000→scalars 

###### [1,0,0,160000,130000,300000]→1D array 

###### [[1,0,0,160000,130000,300000]]→2D array 

###### Important note 2: Notice also that the "California" state was not input as a string in the last column but as "1, 0, 0" in the first three columns. That's because of course the predict method expects the one-hot-encoded values of the state, and as we see in the second row of the matrix of features X, "California" was encoded as "1, 0, 0". And be careful to include these values in the first three columns, not the last three ones, because the dummy variables are always created in the first columns.

Getting the final linear regression equation with the values of the coefficients

In [23]:
print(regressor.coef_)
print(regressor.intercept_)

[ 8.66e+01 -8.73e+02  7.86e+02  7.73e-01  3.29e-02  3.66e-02]
42467.5292485298


Therefore, the equation of our multiple linear regression model is:

Profit=86.6×Dummy State 1−873×Dummy State 2+786×Dummy State 3+0.773×R&D Spend+0.0329×Administration+0.0366×Marketing Spend+42467.53

Important Note: To get these coefficients we called the "coef_" and "intercept_" attributes from our regressor object. Attributes in Python are different than methods and usually return a simple value or an array of values.